docs/designs/supporting_pytorch.md

```diff
@@ -0,0 +1,358 @@
+according to the batch size.
+As we set `BATCH_SIZE = 7`, the result is as follows. The remaining data in this
+epoch will be returned at `step = 1`.
+
+```text
+Epoch:  0 | Step:  0 | batch x:  tensor([1., 7., 2., 8., 9., 6., 3.]) | batch y:  tensor([10.,  4.,  9.,  3.,  2.,  5.,  8.])
+Epoch:  0 | Step:  1 | batch x:  tensor([ 5., 10.,  4.]) | batch y:  tensor([6., 1., 7.])
+Epoch:  1 | Step:  0 | batch x:  tensor([1., 6., 7., 5., 2., 9., 8.]) | batch y:  tensor([10.,  5.,  4.,  6.,  9.,  2.,  3.])
+Epoch:  1 | Step:  1 | batch x:  tensor([10.,  3.,  4.]) | batch y:  tensor([1., 8., 7.])
+Epoch:  2 | Step:  0 | batch x:  tensor([ 4.,  1.,  9.,  5.,  7.,  8., 10.]) | batch y:  tensor([ 7., 10.,  2.,  6.,  4.,  3.,  1.])
+Epoch:  2 | Step:  1 | batch x:  tensor([3., 2., 6.]) | batch y:  tensor([8., 9., 5.])
+```
+
+```python
+import torch
+import torch.utils.data as Data
+
+BATCH_SIZE = 5
+x = torch.linspace(1, 10, 10)       # this is x data (torch tensor)
+y = torch.linspace(10, 1, 10)       # this is y data (torch tensor)
+
+torch_dataset = Data.TensorDataset(x, y)
+loader = Data.DataLoader(
+    dataset=torch_dataset,      # torch TensorDataset format
+    batch_size=BATCH_SIZE,      # mini batch size
+    shuffle=True,               # random shuffle for training
+    num_workers=2,              # subprocesses for loading data
+)
+def show():
+    for epoch in range(3):   # train entire dataset 3 times
+        for step, (batch_x, batch_y) in enumerate(loader):  # for each training step
+            # train your data...
+            print('Epoch: ', epoch, '| Step: ', step, '| batch x: ',
+            batch_x, '| batch y: ', batch_y)
+            # print('Epoch: ', epoch, '| Step: ', step, '| batch x: ',
+            #       batch_x.numpy(), '| batch y: ', batch_y.numpy())
+if __name__ == '__main__':
+    show()
+```
+
+### MNIST Example
+
+This dataset is in numpy array format, and has been stored using pickle, a
+python-specific format for serializing data. Each image is 28 x 28, and is being
+ stored as a flattened row of length 784 (=28x28). Let’s take a look at one; we
+ need to reshape it to 2d first.
+
+```python
+plt.imshow(train_data.train_data[0].numpy(), cmap='gray')
+plt.title('%i' % train_data.train_labels[0])
+plt.show()
+```
+
+PyTorch uses `torch.tensor`, rather than `numpy` arrays.
+
+```python
+x_train, y_train, x_valid, y_valid = map(
+    torch.tensor, (x_train, y_train, x_valid, y_valid)
+)
+```
+
+Or we can convert our data like this.
+
+```python
+train_data = torchvision.datasets.MNIST(
+    root='./mnist/',
+    train=True,                                     # this is training data
+    transform=torchvision.transforms.ToTensor(),    # Converts a PIL.Image or numpy.ndarray to
+                                                    # torch.FloatTensor of shape (C x H x W) and normalize in the range [0.0, 1.0]
+    download=DOWNLOAD_MNIST,
+)
+```
+
+## Training Loop
+
```

Comment:
The training loop should be like this:

```python
while (True):
    task = get_task()
    dataset = create_dataset(task)
    for minibatch in dataset:
        pull_parameters()
        forward()
        backward()
        push_gradients()
```

We have to think about how to implement each step in PyTorch.